src/api/app/models/user.rb

```diff
@@ -859,7 +843,38 @@ def combined_rss_feed_items
   end
 
   def mark_login!
-    update_attributes(last_logged_in_at: Time.now, login_failure_count: 0)
+    update_attributes(last_logged_in_at: Time.zone.today, login_failure_count: 0)
+  end
+
+  def count_login_failure
+    update_attributes(login_failure_count: login_failure_count + 1)
+  end
+
+  def proxy_realname
+    return unless env['HTTP_X_FIRSTNAME'].present? && env['HTTP_X_LASTNAME'].present?
+    env['HTTP_X_FIRSTNAME'].force_encoding('UTF-8') + ' ' + env['HTTP_X_LASTNAME'].force_encoding('UTF-8')
+  end
+
+  def update_login_values(env)
+    # updates user's email and real name using data transmitted by authentication proxy
+    self.email = env['HTTP_X_EMAIL'] if env['HTTP_X_EMAIL'].present?
+    self.realname = proxy_realname if proxy_realname
+
+    self.last_logged_in_at = Time.zone.today
+    self.login_failure_count = 0
+
+    # If the values stay the same after the assignations above, nothing will be updated in database.
+    # So, we won't consider this is a new 'Log in' process.
+    return if changes.blank?
```

Comment:
Isn't that taken care of by `ActiveRecord`?

```ruby
irb(main):002:0> u = User.last
  User Load (0.9ms)  SELECT  `users`.* FROM `users` ORDER BY `users`.`id` DESC LIMIT 1
=> #<User id: 3, created_at: "2019-08-01 10:58:53", updated_at: "2019-08-01 11:09:09", last_logged_in_at: "2019-08-01 10:58:53", login_failure_count: 0, login: "hans", email: "hvogel@hennevogel.de", realname: "", password_digest: nil, deprecated_password: "d7f33cef468756cde99e31f6b8256505", deprecated_password_hash_type: "md5", deprecated_password_salt: nil, adminnote: nil, state: "confirmed", owner_id: nil, ignore_auth_services: false, in_beta: false, in_rollout: true>
irb(main):006:0> u.email               
=> "hvogel@hennevogel.de"
irb(main):007:0> u.email = "hvogel@hennevogel.de"
=> "hvogel@hennevogel.de"
irb(main):008:0> u.save 
   (0.6ms)  BEGIN
  User Exists (0.8ms)  SELECT  1 AS one FROM `users` WHERE `users`.`login` = 'hans' AND `users`.`id` != 3 LIMIT 1
   (0.5ms)  COMMIT
=> true
irb(main):009:0> u.email = "hvogel@hennevogel.com"
=> "hvogel@hennevogel.com"
irb(main):010:0> u.save
   (0.5ms)  BEGIN
  User Exists (0.9ms)  SELECT  1 AS one FROM `users` WHERE `users`.`login` = 'hans' AND `users`.`id` != 3 LIMIT 1
  User Update (6.2ms)  UPDATE `users` SET `updated_at` = '2019-08-15 12:11:22.589734', `email` = 'hvogel@hennevogel.com' WHERE `users`.`id` = 3
   (4.7ms)  COMMIT
=> true
```

No `UPDATE` call in the first save....

Reply:
Yes, in fact I'm based on what you explain when I call save in line 877. That `save` won't perform an UPDATE in database when the value doesn't changed. However, I really need to know (before saving) if there are any changes and only if so I send the message to RabbitMQ and write in the log if email is modified.

I could rewrite this like:

```
  def update_login_values(env)
    # updates user's email and real name using data transmitted by authentication proxy
    self.email = env['HTTP_X_EMAIL'] if env['HTTP_X_EMAIL'].present?
    self.realname = proxy_realname if proxy_realname

    self.last_logged_in_at = Time.zone.today
    self.login_failure_count = 0

    if changes.any?
      # At this point some login value changed, so a successful log in is tracked
      RabbitmqBus.send_to_bus('metrics', 'login,access_point=webui value=1')
    end

    if changes.keys.includes?('email')
      logger.info "updating email for user #{login} from proxy header: old:#{email}|new:#{env['HTTP_X_EMAIL']}"
    end

    save
  end
```

But using `return if changes.blank?` (equivalent to `save && return if changes.blank?` when no UPDATE performed) I avoid going through lines that I'm not going to execute.

Reply:
Forget about the first comment. I have rewritten it and I think is more clear now. Please have a look @hennevogel